src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java

```diff
@@ -0,0 +1,529 @@
+
+/**
+ * This class exposes a method to render a {@code double} as a string.
+ */
+final public class DoubleToDecimal {
+    /*
+     * For full details about this code see the following references:
+     *
+     * [1] Giulietti, "The Schubfach way to render doubles",
+     *     https://drive.google.com/file/d/1gp5xv4CAa78SVgCeWfGqqI4FfYYYuNFb
+     *
+     * [2] IEEE Computer Society, "IEEE Standard for Floating-Point Arithmetic"
+     *
+     * [3] Bouvier & Zimmermann, "Division-Free Binary-to-Decimal Conversion"
+     *
+     * Divisions are avoided altogether for the benefit of those architectures
+     * that do not provide specific machine instructions or where they are slow.
+     * This is discussed in section 10 of [1].
+     */
+
+    /* The precision in bits */
+    static final int P = PRECISION;
+
+    /* Exponent width in bits */
+    private static final int W = (Double.SIZE - 1) - (P - 1);
+
+    /* Minimum value of the exponent: -(2^(W-1)) - P + 3 */
+    static final int Q_MIN = (-1 << (W - 1)) - P + 3;
+
+    /* Maximum value of the exponent: 2^(W-1) - P */
+    static final int Q_MAX = (1 << (W - 1)) - P;
+
+    /* 10^(E_MIN - 1) <= MIN_VALUE < 10^E_MIN */
+    static final int E_MIN = -323;
+
+    /* 10^(E_MAX - 1) <= MAX_VALUE < 10^E_MAX */
+    static final int E_MAX = 309;
+
+    /* Threshold to detect tiny values, as in section 8.2.1 of [1] */
+    static final long C_TINY = 3;
+
+    /* The minimum and maximum k, as in section 8 of [1] */
+    static final int K_MIN = -324;
+    static final int K_MAX = 292;
+
+    /* H is as in section 8.1 of [1] */
+    static final int H = 17;
+
+    /* Minimum value of the significand of a normal value: 2^(P-1) */
+    private static final long C_MIN = 1L << (P - 1);
+
+    /* Mask to extract the biased exponent */
+    private static final int BQ_MASK = (1 << W) - 1;
+
+    /* Mask to extract the fraction bits */
+    private static final long T_MASK = (1L << (P - 1)) - 1;
+
+    /* Used in rop() */
+    private static final long MASK_63 = (1L << 63) - 1;
+
+    /* Used for left-to-tight digit extraction */
+    private static final int MASK_28 = (1 << 28) - 1;
+
+    private static final int NON_SPECIAL    = 0;
+    private static final int PLUS_ZERO      = 1;
+    private static final int MINUS_ZERO     = 2;
+    private static final int PLUS_INF       = 3;
+    private static final int MINUS_INF      = 4;
+    private static final int NAN            = 5;
+
+    /*
+     * Room for the longer of the forms
+     *     -ddddd.dddddddddddd         H + 2 characters
+     *     -0.00ddddddddddddddddd      H + 5 characters
+     *     -d.ddddddddddddddddE-eee    H + 7 characters
+     * where there are H digits d
+     */
+    public static final int MAX_CHARS = H + 7;
+
+    private final byte[] bytes = new byte[MAX_CHARS];
+
+    /* Index into bytes of rightmost valid character */
+    private int index;
+
+    private DoubleToDecimal() {
```

Comment:
Maybe add a comment like
```
    /**
     * Prevent instantiation.
     */
```
or
```
// Prevent instantiation of this class.
```

Reply:
The constructor _is_ invoked, so the comment would be inappropriate.

Reply:
... but I added a `throw` in the constructor of `MathUtils`